src/cargo/ops/cargo_package/vcs.rs

```diff
@@ -263,12 +266,18 @@ fn dirty_files_outside_pkg_root(
 }
 
 /// Helper to collect dirty statuses for a single repo.
-fn collect_statuses(repo: &git2::Repository, dirty_files: &mut Vec<PathBuf>) -> CargoResult<()> {
+fn collect_statuses(
+    repo: &git2::Repository,
+    pathspecs: &[&str],
+    dirty_files: &mut Vec<PathBuf>,
+) -> CargoResult<()> {
```

Comment:
The downside is if your workspace has a real manifest, rather than virtual.  It will end up querying everything.

Collecting statuses once across all packages would help with this (and the outside root files).  This was proposed in #14985 but the real workspace manifest case wasn't considered and it had some drawbacks.